main.tf

```
@@ -70,22 +70,21 @@ resource aws_nat_gateway "nat" {
 #############
 
 resource aws_subnet "public_subnet" {
-  count                   = "${var.az_count}"
+  count                   = "${var.az_count * var.public_subnets_per_az}"
   vpc_id                  = "${aws_vpc.vpc.id}"
   cidr_block              = "${var.public_cidr_ranges[count.index]}"
   availability_zone       = "${element(local.azs, count.index)}"
   map_public_ip_on_launch = true
-  tags                    = "${merge(local.base_tags, map("Name", format("%s-PublicSubnet%d", var.vpc_name, count.index + 1)), var.custom_tags)}"
+  tags                    = "${merge(local.base_tags, map("Name", format("%s-%s%d", var.vpc_name, element(var.public_subnet_names, count.index / var.az_count), count.index + 1)), var.custom_tags)}"
 }
 
 resource aws_subnet "private_subnet" {
-  count                   = "${var.az_count}"
+  count                   = "${var.az_count * var.private_subnets_per_az}"
   vpc_id                  = "${aws_vpc.vpc.id}"
   cidr_block              = "${var.private_cidr_ranges[count.index]}"
   availability_zone       = "${element(local.azs, count.index)}"
   map_public_ip_on_launch = false
-
-  tags = "${merge(local.base_tags, map("Name", format("%s-PrivateSubnet%d", var.vpc_name, count.index + 1)), var.custom_tags)}"
+  tags                    = "${merge(local.base_tags, map("Name", format("%s-%s%d", var.vpc_name, element(var.private_subnet_names, count.index / var.az_count), count.index + 1)), var.custom_tags)}"
```

I think to fix the issue with subnet naming we could update as follows
```suggestion
  format("%s-%s%d", var.vpc_name, element(var.private_subnet_names, count.index / var.az_count), (count.index % var.az_count) + 1))"
```

I would also like to see if we can break up this line.  Its getting a bit long, and if we could break up into multiple lines would help readability.